python/lsst/rapid/analysis/butlerUtils.py

```diff
@@ -0,0 +1,682 @@
+    """Get the day_obs and seq_num for an exposure id.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The butler.
+
+    dataId : `dict`
+        The dataId containing the exposure id.
+
+    Returns
+    -------
+    dataId : `dict`
+        A dict containing only the day_obs and seq_num.
+    """
+    if (dayObs := getDayObs(dataId)) and (seqNum := getSeqNum(dataId)):
+        return {'day_obs': dayObs, 'seq_num': seqNum}
+
+    if isinstance(dataId, int):
+        dataId = {'exposure': dataId}
+    else:
+        dataId = _assureDict(dataId)
+    assert isinstance(dataId, dict)
+
+    if not (expId := getExpId(dataId)):
+        raise RuntimeError(f'Failed to find exposure id in {dataId}')
+
+    where = "exposure.id=expId"
+    expRecords = butler.registry.queryDimensionRecords("exposure",
+                                                       where=where,
+                                                       bind={'expId': expId},
+                                                       datasets='raw')
+    expRecords = list(expRecords)
+    assert len(expRecords) == 1, f'Found more than one exposure record for {dataId}'
+    return {'day_obs': expRecords[0].day_obs, 'seq_num': expRecords[0].seq_num}
+
+
+def getDatasetRefForDataId(butler, datasetType, dataId):
+    """Get the datasetReference for a dataId.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The butler.
+
+    datasetType : `str` or `datasetType`
+        The dataset type.
+
+    dataId : `dict`
+        The dataId.
+
+    Returns
+    -------
+    datasetRef : `lsst.daf.butler.dimensions.DatasetReference`
+        The dataset reference.
+    """
+    if not _expid_present(dataId):
+        assert _dayobs_present(dataId) and _seqnum_present(dataId)
+        dataId.update(getExpIdFromDayObsSeqNum(butler, dataId))
+
+    dRef = butler.registry.findDataset(datasetType, dataId)
+    return dRef
+
+
+def removeDataProduct(butler, datasetType, dataId):
+    """Remove a data prodcut from the registry. Use with caution.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The butler.
+
+    datasetType : `str` or `datasetType`
+        The dataset type.
+
+    dataId : `dict`
+        The dataId.
+
+    """
+    if datasetType == 'raw':
+        raise RuntimeError("I'm sorry, Dave, I'm afraid I can't do that.")
+    dRef = getDatasetRefForDataId(butler, datasetType, dataId)
+    butler.pruneDatasets([dRef], disassociate=True, unstore=True, purge=True)
+    return
+
+
+def _dayobs_present(dataId):
+    return _get_dayobs_key(dataId) is not None
+
+
+def _seqnum_present(dataId):
+    return _get_seqnum_key(dataId) is not None
+
+
+def _expid_present(dataId):
+    return _get_expid_key(dataId) is not None
+
+
+def _get_dayobs_key(dataId):
+    """Return the key for day_obs if present, else None
+    """
+    keys = [k for k, v in dataId.items() if k.find('day_obs') != -1]
```

Comment:
I've changed it to `dataId.keys()` rather than items. I think the original reason was that some butler record thing has a `.items()` method but not a `.keys()`, but the tests still seem to pass and the coverage here is actually pretty good so I'm gonna go with that.

As for the second question: it's the ever-giving joy of `day_obs` and `exposure.day_obs` 🙂